### PR TITLE
Allow the CLI to parse an arguments file

### PR DESCRIPTION
**Pull request by brunoapimentel**
The core implementation is done, working on unit tests and covering corner cases.

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] New code has type annotations
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)
- [ ] [Issue](https://github.com/containerbuildsystem/cachi2/issues/13) that tracks breaking changes to the OSBS2 integration is updated (if applicable)

### CODE REVIEW

File: cachi2/interface/cli.py [hunk 1]
```diff
@@ -95,11 +98,22 @@ def _looks_like_json(value: str) -> bool:
     return value.lstrip().startswith(("{", "["))
 
 
+class ArgsFile(BaseModel, extra="forbid"):
+    packages: list[dict[str, str]] = []
+    flags: list[str] = []
```
Comment (chmeliik) [inline cachi2/interface/cli.py:103]: Maybe this should be an input model?

```python
class SomeArgs(BaseModel, extra="forbid"):
    """Args that you can set in the args file."""
    packages: list[PackageInput]
    flags: list[Flag]


class Request(PackageArgs):
     """All args for a fetch-deps invocation."""
    source_dir: Path
    ...
```

File: pyproject.toml [hunk 2]
```diff
@@ -22,6 +22,7 @@ dependencies = [
     "semver",
     "setuptools",
     "typer",
+    "pyaml",
```
Comment (chmeliik) [inline pyproject.toml:25]: py**y**aml

[pyaml](https://pypi.org/project/pyaml/) is something else - something that apparently produces invalid output as a "feature"
> One good example of such “caveat” is that e.g. {'foo': '123'} will serialize to foo: 123, which for PyYAML would be a bug, as 123 will then be read back as an integer from that, but here it’s a feature.

File: cachi2/core/models/input.py [hunk 3]
```diff
@@ -105,6 +105,12 @@ def check_package_paths(cls, package: PackageInput, values: dict) -> PackageInpu
                 )
         return package
 
+    @pydantic.validator("packages")
+    def non_empty_packages(cls, packages: list[PackageInput]):
+        if len(packages) == 0:
+            raise ValueError("Packages cannot be empty")
+        return packages
```
Comment (chmeliik) [inline cachi2/core/models/input.py:112]: Oof, this will break a lot of tests :smile: 

Maybe this should be validated on the input-parsing side? I don't necessarily think that having 0 packages makes the Request invalid

Comment (chmeliik) [reply cachi2/core/models/input.py:112]: In fact, do we want to validate it at all? Maybe there will be a use case for a package-less cachi2.args file e.g. to make HACBS use network isolation even when no pre-fetch is needed?

File: cachi2/interface/cli.py (hunk 1 above)
Comment (brunoapimentel) [reply cachi2/interface/cli.py:103]: I initially wrote it there... but I think this is very tied to the CLI itself, so I moved it to here at least for now. In the end, this will need to become a valid Request with InputPackages, which sound more generic.

File: pyproject.toml (hunk 2 above)
Comment (brunoapimentel) [reply pyproject.toml:25]: And we see a case of potential typosquatting IRL :face_exhaling:

File: cachi2/core/models/input.py (hunk 3 above)
Comment (brunoapimentel) [reply cachi2/core/models/input.py:112]: My reasoning here was: if the user has an invalid args file, we want to show it to him right away. And probably a good definition of a valid file is one that has minimally one package in it.  If we allow it to be empty, the user could create a valid request by defining the parameters via stdin while having an empty file. 

My main concern here is to leave less room for confusion.  I think it's fine to validade it in the CLI level, though.

+1 for having an extra parameter in case we need packageless requests.

Comment (chmeliik) [reply cachi2/core/models/input.py:112]: So if I'm understanding it correctly, you suggest the validation should fail when the file is empty, even when the user does specify a package via the CLI args?

In that case we would need this validator in ArgsFile, not in Request

File: cachi2/interface/cli.py (hunk 1 above)
Comment (chmeliik) [reply cachi2/interface/cli.py:103]: Hmm, it feels strange to have a "valid" ArgsFile with invalid packages and flags

Comment (brunoapimentel) [reply cachi2/interface/cli.py:103]: Yeah, but that would be caught up in the input model validation, whereas an empty file + stdin packages would not.

Comment (brunoapimentel) [reply cachi2/interface/cli.py:103]: How do we want to proceed with this? I still think the args file is more CLI related than anything else, and the use of Pydantic model here was to:
- fail on extra parameters in the file
- fail in case packages was missing or empty
- the code looks nicer in a class

I'd argue that having a valid ArgsFile with invalid packages is the same as passing the CLI args validation for having a valid JSON, but having it fail in the actual input model validation.

I can move it if you feel like it belongs to the input package, and add real `InputPackages`/`Flags` to the class.

**Comment by brunoapimentel** [comment]
I'm closing this PR because we decided we don't want to have a cachi2.args file living in the root of the repository anymore. Any new efforts in supporting a more generic file parsing will be done in a new PR.
